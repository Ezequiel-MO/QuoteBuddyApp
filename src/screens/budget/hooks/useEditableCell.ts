// src/screens/budget/hooks/useEditableCell.ts
import { useState, useEffect, useRef } from 'react'
import { useUIContext } from '../context/UIContext'

interface UseEditableCellProps {
	value: number
	originalValue?: number
	onSave: (newValue: number) => void
	typeValue?: 'unit' | 'price'
}

export const useEditableCell = ({
	value,
	originalValue,
	onSave,
	typeValue
}: UseEditableCellProps) => {
	// Get the editable context from UIContext
	const { isEditable } = useUIContext()

	// Local state for editing
	const [isEditing, setIsEditing] = useState(false)
	const [localValue, setLocalValue] = useState(value ? value.toString() : '')
	const inputRef = useRef<HTMLInputElement>(null)

	// Update local value when value changes or editing state toggles
	useEffect(() => {
		if (isEditing) {
			inputRef.current?.focus()
		}
		setLocalValue(value ? value.toString() : '')
	}, [isEditing, value])

	// Handle blur (save)
	const handleBlur = () => {
		if (!isEditable) return

		const numericValue = Number(localValue)
		if (!isNaN(numericValue) && numericValue !== value) {
			onSave(numericValue)
		}
		setIsEditing(false)
	}

	// Handle input change
	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		setLocalValue(e.target.value)
	}

	// Handle key down (for Enter and Escape)
	const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
		if (e.key === 'Enter') {
			handleBlur()
		}
		if (e.key === 'Escape') {
			setIsEditing(false)
			setLocalValue(value.toString())
		}
	}

	// Handle click to start editing
	const handleClick = () => {
		if (!isEditable) return
		setIsEditing(true)
	}

	// Check if value has been modified
	const isModified =
		typeof originalValue === 'number' && originalValue !== value

	return {
		isEditing,
		localValue,
		inputRef,
		isModified,
		isEditable,
		handleBlur,
		handleChange,
		handleKeyDown,
		handleClick
	}
}
