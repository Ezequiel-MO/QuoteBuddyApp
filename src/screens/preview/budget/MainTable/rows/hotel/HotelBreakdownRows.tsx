import { HotelBreakdownRow } from '.'
import { Icon } from '@iconify/react'
import { useContextBudget } from '../../../context/BudgetContext'

interface Props {
  isOpen: boolean
}

export const HotelBreakdownRows = ({ isOpen }: Props) => {
  const { state } = useContextBudget()
  if (!state.selectedHotel) return null

  const {
    DUInr = 0,
    DUIprice = 0,
    DoubleRoomNr = 0,
    DoubleRoomPrice = 0,
    DailyTax = 0,
    breakfast = 0
  } = state.selectedHotel.price[0]

  return (
    <>
      <tr
        style={{
          transition: 'all 0.5s ease-in-out',
          maxHeight: isOpen ? '800px' : '0',
          overflow: 'hidden',
          opacity: isOpen ? '1' : '0'
        }}
      >
        <td colSpan={6} className='p-0 bg-transparent'>
          <div
            className={`transition-all duration-500 ease-in-out overflow-hidden ${
              isOpen ? 'max-h-[800px] opacity-100' : 'max-h-0 opacity-0'
            }`}
          >
            <table className='w-full'>
              <tbody className='w-full bg-white-100 dark:bg-[#a9ba9d] relative'>
                <tr>
                  <td colSpan={6} className='p-0 bg-transparent'>
                    <div className='absolute inset-0 flex items-center justify-center opacity-10 dark:opacity-20 z-0'>
                      <Icon icon='ic:twotone-local-hotel' width={300} />
                    </div>
                    <table className='w-full'>
                      <thead className='text-white-100 bg-zinc-800'>
                        <tr>
                          <td align='center'>Description</td>
                          <td align='center'>Nr. Units </td>
                          <td align='center'>Nr. of nights </td>
                          <td align='center'>Cost per room per night</td>
                          <td align='center'>Total Cost</td>
                        </tr>
                      </thead>
                      <tbody className='text-[#000]'>
                        <HotelBreakdownRow
                          units={DUInr}
                          rate={DUIprice}
                          nights={state.schedule.length - 1}
                          title='Double Room Single Use'
                        />
                        <HotelBreakdownRow
                          units={DoubleRoomNr}
                          rate={DoubleRoomPrice}
                          nights={state.schedule.length - 1}
                          title='Double Room //Twin Room'
                        />
                        <HotelBreakdownRow
                          units={DUInr + DoubleRoomNr * 2}
                          rate={DailyTax}
                          nights={state.schedule.length - 1}
                          title='City Tax'
                        />
                        {breakfast ? (
                          <HotelBreakdownRow
                            units={DUInr + DoubleRoomNr * 2}
                            rate={breakfast}
                            nights={state.schedule.length - 1}
                            title='Breakfast'
                          />
                        ) : null}
                      </tbody>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </td>
      </tr>
    </>
  )
}
