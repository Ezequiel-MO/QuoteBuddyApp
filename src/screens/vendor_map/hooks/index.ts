export * from './useMapOptions'
export * from './useMapZoom'
export * from './useVendorCoords'
