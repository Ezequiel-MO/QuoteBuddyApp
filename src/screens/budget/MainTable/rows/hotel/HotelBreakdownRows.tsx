import React, { useEffect, useState } from 'react'
import { HotelBreakdownRow } from '.'
import { useCurrentProject } from 'src/hooks'
import { Spinner } from 'src/components/atoms/spinner/Spinner'
import { IDay } from '@interfaces/project'

interface Props {
	isOpen: boolean
}

export const HotelBreakdownRows: React.FC<Props> = ({ isOpen }) => {
	const [isLoading, setIsLoading] = useState(false)
	const { currentProject, budget } = useCurrentProject()

	const selectedHotel = budget.selectedHotel
	const schedule: IDay[] = currentProject.schedule || []

	useEffect(() => {
		setIsLoading(true)
		setTimeout(() => {
			setIsLoading(false)
		}, 1000)
	}, [selectedHotel?._id])

	if (!selectedHotel) {
		return null
	}

	const { price } = selectedHotel
	if (!price || price.length === 0) {
		return null
	}

	const {
		DUInr = 0,
		DUIprice = 0,
		DoubleRoomNr = 0,
		DoubleRoomPrice = 0,
		DailyTax = 0,
		breakfast = 0
	} = price[0]

	const numberOfNights = schedule.length - 1

	const renderBreakdownRows = () => (
		<>
			<HotelBreakdownRow
				units={DUInr}
				rate={DUIprice}
				nights={numberOfNights}
				title="Double Room Single Use"
			/>
			<HotelBreakdownRow
				units={DoubleRoomNr}
				rate={DoubleRoomPrice}
				nights={numberOfNights}
				title="Double Room // Twin Room"
			/>
			<HotelBreakdownRow
				units={DUInr + DoubleRoomNr * 2}
				rate={DailyTax}
				nights={numberOfNights}
				title="City Tax"
			/>
			{breakfast > 0 && (
				<HotelBreakdownRow
					units={DUInr + DoubleRoomNr * 2}
					rate={breakfast}
					nights={numberOfNights}
					title="Breakfast"
				/>
			)}
		</>
	)

	return (
		<tr>
			<td colSpan={6} className="p-0 bg-transparent">
				<div
					className={`transition-all duration-500 ease-in-out overflow-hidden ${
						isOpen ? 'max-h-[800px] opacity-100' : 'max-h-0 opacity-0'
					}`}
				>
					<table className="w-full">
						<tbody className="w-full bg-white-100 dark:bg-[#a9ba9d] relative">
							<tr>
								<td colSpan={6} className="p-0 bg-transparent">
									{isLoading ? (
										<div className="flex items-center justify-center  dark:opacity-20 z-0 pointer-events-none">
											<Spinner aria-label="Loading hotel details" />
										</div>
									) : (
										<table className="w-full">
											<thead className="text-white-100 bg-zinc-800">
												<tr>
													<td align="center">Description</td>
													<td align="center">Nr. Units</td>
													<td align="center">Nr. of Nights</td>
													<td align="center">Cost per Room per Night</td>
													<td align="center">Total Cost</td>
												</tr>
											</thead>
											<tbody className="text-[#000]">
												{renderBreakdownRows()}
											</tbody>
										</table>
									)}
								</td>
							</tr>
						</tbody>
					</table>
				</div>
			</td>
		</tr>
	)
}
