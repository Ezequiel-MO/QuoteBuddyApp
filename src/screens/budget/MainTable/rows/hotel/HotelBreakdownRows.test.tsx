import {
	describe,
	it,
	beforeEach,
	afterEach,
	expect,
	vi,
	type Mock
} from 'vitest'
import { useCurrentProject } from 'src/hooks'
import { starterHotel, starterSchedule } from '@constants/starterObjects'
import { IDay } from '@interfaces/project'
import { HotelBreakdownRows } from './HotelBreakdownRows'
import { act, render, screen, within } from '@testing-library/react'

vi.mock('./HotelBreakdownRow', () => ({
	HotelBreakdownRow: ({
		units,
		rate,
		nights,
		title
	}: {
		units: number
		rate: number
		nights: number
		title: string
	}) => (
		<tr data-testid="HotelBreakdownRow">
			<td>{title}</td>
			<td>{units}</td>
			<td>{nights}</td>
			<td>{rate}</td>
			<td>{units * rate * nights}</td>
		</tr>
	)
}))

vi.mock('src/components/atoms/spinner/Spinner', () => ({
	Spinner: () => <div data-testid="Spinner">Loading...</div>
}))

vi.mock('src/hooks', () => ({
	useCurrentProject: vi.fn()
}))

const mockUseCurrentProject = useCurrentProject as Mock

describe('HotelBreakdownRows', () => {
	const mockSetBudgetSelectedHotel = vi.fn()
	const mockSetBudgetSelectedHotelCost = vi.fn()
	const mockUpdateBudgetMeetingsTotalCost = vi.fn()
	const mockClearMeetingsBudget = vi.fn()

	const mockHotel1 = {
		...starterHotel,
		name: 'Hotel One',
		price: [
			{
				DUInr: 2,
				DUIprice: 100,
				DoubleRoomNr: 1,
				DoubleRoomPrice: 150,
				DailyTax: 20,
				breakfast: 10
			}
		]
	}

	const mockHotel2 = {
		...starterHotel,
		name: 'Hotel Two',
		price: [
			{
				DUInr: 1,
				DUIprice: 80,
				DoubleRoomNr: 2,
				DoubleRoomPrice: 120,
				DailyTax: 15,
				breakfast: 10
			}
		]
	}

	const mockSchedule: IDay[] = starterSchedule

	beforeEach(() => {
		vi.useFakeTimers()
		vi.clearAllMocks()
		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotel1, mockHotel2],
				schedule: mockSchedule
			},
			budget: {
				selectedHotel: mockHotel1
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})
	})

	afterEach(() => {
		vi.useRealTimers()
		vi.restoreAllMocks()
	})

	const renderComponent = (props: { isOpen: boolean } = { isOpen: false }) =>
		render(
			<table>
				<tbody>
					<HotelBreakdownRows {...props} />
				</tbody>
			</table>
		)

	it('renders nothing when selectedHotel is null', () => {
		const mockHotelWithoutPrice = {
			...starterHotel,
			name: 'Hotel Three',
			price: []
		}
		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotelWithoutPrice],
				schedule: mockSchedule
			},
			budget: {
				selectedHotel: mockHotelWithoutPrice
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		const { container } = renderComponent({ isOpen: true })
		expect(container.querySelector('tr')).toBeNull()
	})

	it('displays spinner while loading', async () => {
		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		expect(screen.getByTestId('Spinner')).toBeInTheDocument()

		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		expect(screen.queryByTestId('Spinner')).not.toBeInTheDocument()

		vi.useRealTimers()
	})

	it('toggles visibility based on isOpen prop', async () => {
		vi.useFakeTimers()
		const { rerender } = renderComponent({ isOpen: false })
		const visibilityContainer = screen.getByTestId('visibility-container')
		expect(visibilityContainer).toHaveClass('max-h-0 opacity-0')
		await act(async () => {
			vi.advanceTimersByTime(1000)
		})
		rerender(
			<table>
				<tbody>
					<HotelBreakdownRows isOpen={true} />
				</tbody>
			</table>
		)

		expect(visibilityContainer).toHaveClass('max-h-[800px] opacity-100')
		vi.useRealTimers()
	})

	it('renders 4 HotelBreakdownRow components when breakfast > 0', async () => {
		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		await act(async () => {
			vi.advanceTimersByTime(1000)
			const rows = screen.getAllByTestId('HotelBreakdownRow')
			expect(rows.length).toBe(4)
			expect(rows[3]).toHaveTextContent('Breakfast')
		})

		vi.useRealTimers()
	})

	it('renders 3 HotelBreakdownRow components when breakfast = 0', async () => {
		const mockHotelNoBreakfast = {
			...starterHotel,
			name: 'Hotel No Breakfast',
			price: [
				{
					DUInr: 2,
					DUIprice: 100,
					DoubleRoomNr: 1,
					DoubleRoomPrice: 150,
					DailyTax: 20,
					breakfast: 0 // Set breakfast to 0
				}
			]
		}

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotelNoBreakfast, mockHotel2],
				schedule: mockSchedule
			},
			budget: {
				selectedHotel: mockHotelNoBreakfast
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		const rows = screen.getAllByTestId('HotelBreakdownRow')
		expect(rows.length).toBe(3)
		expect(rows[2]).not.toHaveTextContent('Breakfast')

		vi.useRealTimers()
	})

	it('calculates numberOfNights correctly for schedule with 2 days', async () => {
		const mockScheduleTwoDays: IDay[] = [{} as IDay, {} as IDay]

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotel1, mockHotel2],
				schedule: mockScheduleTwoDays
			},
			budget: {
				selectedHotel: mockHotel1
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		await act(async () => {
			vi.advanceTimersByTime(1000)
		})
		const rows = screen.getAllByTestId('HotelBreakdownRow')
		const cityTaxRow = rows.find((row) => row.textContent?.includes('City Tax'))
		expect(cityTaxRow).toBeDefined()

		if (cityTaxRow) {
			const { getAllByRole } = within(cityTaxRow)
			const cells = getAllByRole('cell')

			expect(cells[2]).toHaveTextContent('1')
		}

		vi.useRealTimers()
	})

	it('calculates numberOfNights correctly for schedule with 1 day', async () => {
		const mockScheduleOneDay: IDay[] = [{} as IDay]

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotel1, mockHotel2],
				schedule: mockScheduleOneDay
			},
			budget: {
				selectedHotel: mockHotel1
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		// Fast-forward the timeout to end loading
		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		const rows = screen.getAllByTestId('HotelBreakdownRow')
		const cityTaxRow = rows.find((row) => row.textContent?.includes('City Tax'))
		expect(cityTaxRow).toBeDefined()

		if (cityTaxRow) {
			const { getAllByRole } = within(cityTaxRow)
			const cells = getAllByRole('cell')
			// cells[2] is the 'Nr. of Nights' column
			expect(cells[2]).toHaveTextContent('0')
		}

		vi.useRealTimers()
	})

	it('handles empty schedule gracefully by setting numberOfNights to 0', async () => {
		const mockScheduleEmpty: IDay[] = []

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotel1, mockHotel2],
				schedule: mockScheduleEmpty
			},
			budget: {
				selectedHotel: mockHotel1
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		const rows = screen.getAllByTestId('HotelBreakdownRow')
		const cityTaxRow = rows.find((row) => row.textContent?.includes('City Tax'))
		expect(cityTaxRow).toBeDefined()

		if (cityTaxRow) {
			const { getAllByRole } = within(cityTaxRow)
			const cells = getAllByRole('cell')
			expect(cells[2]).toHaveTextContent('0')
		}

		vi.useRealTimers()
	})

	it('renders correct HotelBreakdownRow components for a different selectedHotel', async () => {
		const mockHotelSelected = mockHotel2

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [mockHotel1, mockHotel2],
				schedule: mockSchedule // Ensure schedule has the expected length
			},
			budget: {
				selectedHotel: mockHotelSelected
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		vi.useFakeTimers()
		renderComponent({ isOpen: true })

		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		const rows = screen.getAllByTestId('HotelBreakdownRow')
		expect(rows.length).toBe(4)
		expect(rows[0]).toHaveTextContent('Double Room Single Use')
		expect(rows[1]).toHaveTextContent('Double Room // Twin Room')
		expect(rows[2]).toHaveTextContent('City Tax')
		expect(rows[3]).toHaveTextContent('Breakfast')

		// First Row: Double Room Single Use
		const firstRow = rows[0]
		const { getAllByRole: getAllByRoleFirstRow } = within(firstRow)
		const cellsFirstRow = getAllByRoleFirstRow('cell')
		expect(cellsFirstRow[1]).toHaveTextContent('1') // DUInr
		expect(cellsFirstRow[3]).toHaveTextContent('80') // DUIprice
		expect(cellsFirstRow[2]).toHaveTextContent('1') // Nights

		// Second Row: Double Room // Twin Room
		const secondRow = rows[1]
		const { getAllByRole: getAllByRoleSecondRow } = within(secondRow)
		const cellsSecondRow = getAllByRoleSecondRow('cell')
		expect(cellsSecondRow[1]).toHaveTextContent('2') // DoubleRoomNr
		expect(cellsSecondRow[3]).toHaveTextContent('120') // DoubleRoomPrice
		expect(cellsSecondRow[2]).toHaveTextContent('1') // Nights

		// Third Row: City Tax
		const cityTaxRow = rows[2]
		const { getAllByRole: getAllByRoleCityTaxRow } = within(cityTaxRow)
		const cellsCityTaxRow = getAllByRoleCityTaxRow('cell')
		expect(cellsCityTaxRow[1]).toHaveTextContent('5') // DUInr + DoubleRoomNr * 2 =1 +2*2=5
		expect(cellsCityTaxRow[3]).toHaveTextContent('15') // DailyTax
		expect(cellsCityTaxRow[2]).toHaveTextContent('1') // Nights

		// Fourth Row: Breakfast
		const breakfastRow = rows[3]
		const { getAllByRole: getAllByRoleBreakfastRow } = within(breakfastRow)
		const cellsBreakfastRow = getAllByRoleBreakfastRow('cell')
		expect(cellsBreakfastRow[1]).toHaveTextContent('5') // DUInr + DoubleRoomNr * 2 =1 +2*2=5
		expect(cellsBreakfastRow[3]).toHaveTextContent('10') // breakfast
		expect(cellsBreakfastRow[2]).toHaveTextContent('1') // Nights

		vi.useRealTimers()
	})

	it('handles error thrown by useCurrentProject gracefully', () => {
		// Mock useCurrentProject to throw an error
		mockUseCurrentProject.mockImplementation(() => {
			throw new Error('useCurrentProject hook error')
		})

		// Spy on console.error to suppress error logs during test
		const consoleErrorSpy = vi
			.spyOn(console, 'error')
			.mockImplementation(() => {})

		expect(() => renderComponent({ isOpen: true })).toThrow(
			'useCurrentProject hook error'
		)

		// Restore console.error
		consoleErrorSpy.mockRestore()
	})

	it('updates dependentUnits correctly when DUInr is edited', async () => {
		// Initial state: DUInr = 2, DoubleRoomNr = 1, dependentUnits = 2 + 2*1 = 4
		const { rerender } = renderComponent({ isOpen: true })

		// Advance timers to end loading
		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		// Find the "City Tax" row and verify dependentUnits
		const cityTaxRow = screen.getByText('City Tax').closest('tr')
		expect(cityTaxRow).toBeInTheDocument()
		if (cityTaxRow) {
			const { getAllByRole } = within(cityTaxRow)
			const cells = getAllByRole('cell')
			expect(cells[1]).toHaveTextContent('4') // dependentUnits = 4
		}

		// Update DUInr to 3: new dependentUnits = 3 + 2*1 = 5
		const updatedHotel1 = {
			...mockHotel1,
			price: [
				{
					...mockHotel1.price[0],
					DUInr: 3 // Edited DUInr
				}
			]
		}

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [updatedHotel1, mockHotel2],
				schedule: mockSchedule
			},
			budget: {
				selectedHotel: updatedHotel1
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		// Re-render the component with updated mock
		rerender(
			<table>
				<tbody>
					<HotelBreakdownRows isOpen={true} />
				</tbody>
			</table>
		)

		// Advance timers to end loading after re-render
		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		// Verify that "City Tax" row has updated dependentUnits
		const updatedCityTaxRow = screen.getByText('City Tax').closest('tr')
		expect(updatedCityTaxRow).toBeInTheDocument()
		if (updatedCityTaxRow) {
			const { getAllByRole } = within(updatedCityTaxRow)
			const cells = getAllByRole('cell')
			expect(cells[1]).toHaveTextContent('5') // dependentUnits = 5
		}
	})

	it('updates dependentUnits correctly when DoubleRoomNr is edited', async () => {
		// Initial state: DUInr = 2, DoubleRoomNr = 1, dependentUnits = 2 + 2*1 = 4
		const { rerender } = renderComponent({ isOpen: true })

		// Advance timers to end loading
		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		// Find the "City Tax" row and verify dependentUnits
		const cityTaxRow = screen.getByText('City Tax').closest('tr')
		expect(cityTaxRow).toBeInTheDocument()
		if (cityTaxRow) {
			const { getAllByRole } = within(cityTaxRow)
			const cells = getAllByRole('cell')
			expect(cells[1]).toHaveTextContent('4') // dependentUnits = 4
		}

		// Update DoubleRoomNr to 2: new dependentUnits = 2 + 2*2 = 6
		const updatedHotel1 = {
			...mockHotel1,
			price: [
				{
					...mockHotel1.price[0],
					DoubleRoomNr: 2 // Edited DoubleRoomNr
				}
			]
		}

		mockUseCurrentProject.mockReturnValue({
			currentProject: {
				multiDestination: false,
				hotels: [updatedHotel1, mockHotel2],
				schedule: mockSchedule
			},
			budget: {
				selectedHotel: updatedHotel1
			},
			setBudgetSelectedHotel: mockSetBudgetSelectedHotel,
			setBudgetSelectedHotelCost: mockSetBudgetSelectedHotelCost,
			updateBudgetMeetingsTotalCost: mockUpdateBudgetMeetingsTotalCost,
			clearMeetingsBudget: mockClearMeetingsBudget
		})

		// Re-render the component with updated mock
		rerender(
			<table>
				<tbody>
					<HotelBreakdownRows isOpen={true} />
				</tbody>
			</table>
		)

		// Advance timers to end loading after re-render
		await act(async () => {
			vi.advanceTimersByTime(1000)
		})

		// Verify that "City Tax" row has updated dependentUnits
		const updatedCityTaxRow = screen.getByText('City Tax').closest('tr')
		expect(updatedCityTaxRow).toBeInTheDocument()
		if (updatedCityTaxRow) {
			const { getAllByRole } = within(updatedCityTaxRow)
			const cells = getAllByRole('cell')
			expect(cells[1]).toHaveTextContent('6') // dependentUnits = 6
		}
	})
})
