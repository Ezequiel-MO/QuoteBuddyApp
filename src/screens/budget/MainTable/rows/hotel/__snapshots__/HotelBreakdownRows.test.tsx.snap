// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelBreakdownRows > matches snapshot when loaded and isOpen = false 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="p-0 bg-transparent"
          colspan="6"
        >
          <div
            class="transition-all duration-500 ease-in-out overflow-hidden max-h-0 opacity-0"
            data-testid="visibility-container"
          >
            <table
              class="w-full"
            >
              <tbody
                class="w-full bg-white-100 dark:bg-[#a9ba9d] relative"
              >
                <tr>
                  <td
                    class="p-0 bg-transparent"
                    colspan="6"
                  >
                    <table
                      class="w-full"
                    >
                      <thead
                        class="text-white-100 bg-zinc-800"
                      >
                        <tr>
                          <td
                            align="center"
                          >
                            Description
                          </td>
                          <td
                            align="center"
                          >
                            Nr. Units
                          </td>
                          <td
                            align="center"
                          >
                            Nr. of Nights
                          </td>
                          <td
                            align="center"
                          >
                            Cost per Room per Night
                          </td>
                          <td
                            align="center"
                          >
                            Total Cost
                          </td>
                        </tr>
                      </thead>
                      <tbody
                        class="text-[#000]"
                      >
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            Double Room Single Use
                          </td>
                          <td>
                            2
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            100
                          </td>
                          <td>
                            200
                          </td>
                        </tr>
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            Double Room // Twin Room
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            150
                          </td>
                          <td>
                            150
                          </td>
                        </tr>
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            City Tax
                          </td>
                          <td>
                            4
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            20
                          </td>
                          <td>
                            80
                          </td>
                        </tr>
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            Breakfast
                          </td>
                          <td>
                            4
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            10
                          </td>
                          <td>
                            40
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`HotelBreakdownRows > matches snapshot when loaded and isOpen = true 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="p-0 bg-transparent"
          colspan="6"
        >
          <div
            class="transition-all duration-500 ease-in-out overflow-hidden max-h-[800px] opacity-100"
            data-testid="visibility-container"
          >
            <table
              class="w-full"
            >
              <tbody
                class="w-full bg-white-100 dark:bg-[#a9ba9d] relative"
              >
                <tr>
                  <td
                    class="p-0 bg-transparent"
                    colspan="6"
                  >
                    <table
                      class="w-full"
                    >
                      <thead
                        class="text-white-100 bg-zinc-800"
                      >
                        <tr>
                          <td
                            align="center"
                          >
                            Description
                          </td>
                          <td
                            align="center"
                          >
                            Nr. Units
                          </td>
                          <td
                            align="center"
                          >
                            Nr. of Nights
                          </td>
                          <td
                            align="center"
                          >
                            Cost per Room per Night
                          </td>
                          <td
                            align="center"
                          >
                            Total Cost
                          </td>
                        </tr>
                      </thead>
                      <tbody
                        class="text-[#000]"
                      >
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            Double Room Single Use
                          </td>
                          <td>
                            2
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            100
                          </td>
                          <td>
                            200
                          </td>
                        </tr>
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            Double Room // Twin Room
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            150
                          </td>
                          <td>
                            150
                          </td>
                        </tr>
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            City Tax
                          </td>
                          <td>
                            4
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            20
                          </td>
                          <td>
                            80
                          </td>
                        </tr>
                        <tr
                          data-testid="HotelBreakdownRow"
                        >
                          <td>
                            Breakfast
                          </td>
                          <td>
                            4
                          </td>
                          <td>
                            1
                          </td>
                          <td>
                            10
                          </td>
                          <td>
                            40
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`HotelBreakdownRows > matches snapshot when loading 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="p-0 bg-transparent"
          colspan="6"
        >
          <div
            class="transition-all duration-500 ease-in-out overflow-hidden max-h-[800px] opacity-100"
            data-testid="visibility-container"
          >
            <table
              class="w-full"
            >
              <tbody
                class="w-full bg-white-100 dark:bg-[#a9ba9d] relative"
              >
                <tr>
                  <td
                    class="p-0 bg-transparent"
                    colspan="6"
                  >
                    <div
                      class="flex items-center justify-center  dark:opacity-20 z-0 pointer-events-none"
                    >
                      <div
                        data-testid="Spinner"
                      >
                        Loading...
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
